pkg/upgrade/upgrades/v25_1_add_jobs_tables.go

```diff
@@ -95,57 +94,76 @@ func backfillJobsTablesAndColumns(
 	log.Infof(ctx, "backfilling new jobs tables and columns")
 	jobsBackfilled := 0
 	for {
+		candidateRows, err := d.DB.Executor().QueryBufferedEx(ctx, "jobs-backfill-find", nil, sessiondata.NodeUserSessionDataOverride,
+			`SELECT id FROM system.jobs WHERE owner IS NULL LIMIT $1`, jobsBackfillPageSize)
+		if err != nil {
+			return err
+		}
+		if len(candidateRows) == 0 {
+			// All done!
+			break
+		}
+
+		candidates := make([]int, len(candidateRows))
+		for i := range candidateRows {
+			candidates[i] = int(tree.MustBeDInt(candidateRows[i][0]))
+		}
 		var done int
 		if err := d.DB.DescsTxn(ctx, func(ctx context.Context, tx descs.Txn) (retErr error) {
-			// Find rows that have not been backfilled, which we will detect via a
-			// NULL owner. The coalesce() here isn't necessary since the vtable will
-			// produce empty strings rather than nulls if a job somehow is missing an
-			// owner, but since the owner not being null is what breaks the loop the
-			// extra coalesce here just ensures wouldn't loop forever if that were to
-			// change out from under us.
-			q, err := tx.QueryIteratorEx(ctx, "jobs-backfill-read", tx.KV(), sessiondata.NodeUserSessionDataOverride,
-				`SELECT
-				v.job_id,
-				v.description,
-				coalesce(v.user_name, ''),
-				v.finished,
-				v.error,
-				v.running_status,
-				v.fraction_completed,
-				v.high_water_timestamp
-				FROM crdb_internal.jobs v
-				LEFT JOIN system.jobs j ON j.id = v.job_id
-				WHERE j.owner IS NULL
-				LIMIT $1`, jobsBackfillPageSize,
-			)
+			if _, err := tx.ExecEx(ctx, "jobs-backfill-upgrade-locking", tx.KV(), sessiondata.NodeUserSessionDataOverride,
+				"SET enable_durable_locking_for_serializable = true"); err != nil {
+				return err
+			}
+
+			// re-read the jobs in the txn, acquiring locks this time.
+			todoRows, err := tx.QueryBufferedEx(ctx, "jobs-backfill-lock-job", tx.KV(), sessiondata.NodeUserSessionDataOverride,
+				`SELECT id FROM system.jobs WHERE id = ANY($1) AND owner IS NULL FOR UPDATE`, candidates)
 			if err != nil {
 				return err
 			}
-			defer func() {
-				retErr = errors.CombineErrors(retErr, q.Close())
-			}()
 
-			done = 0
-			for {
-				ok, err := q.Next(ctx)
+			// No candidates are remained eligible as of txn; move on.
+			if len(todoRows) == 0 {
+				return nil
+			}
+
+			todo := candidates[:0]
+			for i := range todoRows {
+				todo = append(todo, int(tree.MustBeDInt(todoRows[i][0])))
+			}
+			for _, id := range candidates {
+				// Lock the job infos to prevent concurrent modifications there as well.
+				_, err = tx.QueryBufferedEx(ctx, "jobs-backfill-lock-info", tx.KV(), sessiondata.NodeUserSessionDataOverride,
+					`SELECT job_id FROM system.job_info WHERE job_id = ANY($1) FOR UPDATE`, todo)
```

Comment:
nit: this ANY operator is unnecessary. hopefully the optimizer knows this :)

Reply:
oh, this stmt should be above the loop.

Reply:
note it is passing all of `todo` not just `id`, so we only need to run it once, before the loop (the extra runs are a slight waste but don't affect the correctness so I don't think moving it up invalidates our testing so far).

Reply:
oh yeah. good stress the system a little more :)

Reply:
Okay, gone now -- I ended up moving the txn to be one per job in the end since I was still seeing (unexplained) read/write conflicts here even with reading just one row at a time from the vtable in the txn. Moving to one txn per job seems to be making progress even in the fact of my extreme (i.e. far worse than the adoption loop can handle) stress test.